lib/screens/v2/candidate_generator/dup.ex

```diff
@@ -115,86 +135,234 @@ defmodule Screens.V2.CandidateGenerator.Dup do
-      if Enum.any?(secondary_sections_data, &(&1 == :error)) do
-        %DeparturesNoData{screen: config, show_alternatives?: true}
-      else
-        sections =
-          Enum.map(secondary_sections_data, fn {:ok, departures} ->
-            visible_departures =
-              if length(secondary_sections_data) > 1 do
-                Enum.take(departures, 2)
-              else
-                Enum.take(departures, 4)
-              end
-
-            %{type: :normal_section, rows: visible_departures}
-          end)
-
-        [
-          %DeparturesWidget{
-            screen: config,
-            section_data: sections,
-            slot_names: [:main_content_two]
-          }
-        ]
-      end
+      sections_data_to_departure_instances(
+        config,
+        secondary_sections_data,
+        [:main_content_two],
+        now
+      )
 
     primary_departures_instances ++ secondary_departures_instances
   end
 
+  defp sections_data_to_departure_instances(config, sections_data, slot_ids, now) do
+    if Enum.any?(sections_data, &(&1 == :error)) do
+      %DeparturesNoData{screen: config, show_alternatives?: true}
+    else
+      sections =
+        Enum.map(sections_data, fn %{
+                                     departures: departures,
+                                     pill: pill,
+                                     alert: alert,
+                                     headway: headway,
+                                     stop_ids: stop_ids
+                                   } ->
+          visible_departures =
+            if length(sections_data) > 1 do
+              Enum.take(departures, 2)
+            else
+              Enum.take(departures, 4)
+            end
+
+          case get_headway_mode(stop_ids, headway, alert, now) do
+            {:active, time_range, headsign} ->
+              %{type: :headway_section, pill: pill, time_range: time_range, headsign: headsign}
+
+            :inactive ->
+              %{type: :normal_section, rows: visible_departures}
+          end
+        end)
+
+      Enum.map(slot_ids, fn slot_id ->
+        %DeparturesWidget{
+          screen: config,
+          section_data: sections,
+          slot_names: [slot_id]
+        }
+      end)
+    end
+  end
+
+  defp get_sections_data(sections, fetch_section_departures_fn, fetch_alerts_fn) do
+    sections
+    |> Task.async_stream(fn %Section{
+                              query: %Query{
+                                params: %Params{stop_ids: stop_ids, route_ids: route_ids} = params
+                              },
+                              headway: headway
+                            } = section ->
+      section_departures =
+        case fetch_section_departures_fn.(section) do
+          {:ok, section_departures} -> section_departures
+          _ -> []
+        end
+
+      section_route = get_section_route(route_ids)
+
+      section_alert = get_section_alert(params, section_route, fetch_alerts_fn)
```

Comment:
This assumes that all subway/light-rail departure sections are configured with `route_ids`. That is not currently the case for Broadway, for example. 

However! I do think a consistent configuration pattern _should_ be something we do. What would the ideal configuration structure be? Always use stop_ids for bus, but route_ids for everything else? Omit route_type from the config completely?

Reply:
On the other hand, that seems unnecessarily complicated, because what if people don't remember this rule? Maybe we should permit both `route_ids` and `stop_ids`, but if `route_ids` is blank, then we should get the routes from stop_ids before fetching alerts

Reply:
I agree, mandatory config values for minor things like pills is risky. I tweaked the logic a bit so it uses an existing function that was created for the ReconstructedAlert widget. It is a sort of internal caching function that will get the routes for a stop from the cache if it exists. If it doesn't, it fetches them from the API and then saves them.